Deluge 2.0 Support

bdutro opened the pull request:
Fix for issue #36

bdutro committed:
- Convert to Deluge 2.0

bdutro committed:
- Fix some deprecation warnings

bdutro committed:
- Fix RPC methods to return rencode-able types
- Bump version to avoid conflicts with already installed versions

bdutro committed:
- Serialize datetime.datetime objects to dict instead of isoformat to
preserve Python 3.6 compatibility

perfect7gentleman commented:
thanx. testing.

mhertz commented:
Thank you so much bdutro, works for me in my testing and posted about on deluge-forum! :)

Sorry for off-topic, I previously made a patch for previous labelplus and deluge(before v2) to fix/workaround auto-labeling not working for magnets, but this fix/workaround doesn't work anymore. have you an idea of where to place the sleep() call in your version, or maybe it's because of a deluge change instead, as the wait does happen I can see after magnets are added, but still doesn't work, and I tried adding a sleep() call in between all the lines in that section without it working, and even used 10 secs. Thanks again regardless! :)
https://gist.github.com/mhertz/d9c91befb4f255409e8a233fb0e2431f (The patch uses 1 secs, though I've changed to 3 secs recently to be more sure)

bdutro commented:
@mhertz, how are you adding the magnet links? Is it just through the add torrent dialog?

Edit: It appears to be working for me locally when I add a magnet link from the dialog.

mhertz commented:
Ohh, sorry, I remember Cas adding pre-fetching metadata for magnets in GTK-UI some time ago, so it works there seemingly(from your testing), but I always test from terminal/script with 'deluge-console add foo', and that didn't work without adding a slight delay, but which doesn't work anymore. I don't know if it's because of the added lines/things you added to that section, compared to last labelplus, or, a deluge2 change altogether. I'm testing the usecase of using auto-labeling to matching of filename-extensions of any file in torrent/magnet, like (*.mkv|*.mp4|*.avi) auto-moved to e.g. '~/Video' etc.

bdutro commented:
It looks like there is a bug in 2.0.3 that keeps the file metadata from getting prefetched properly:
https://forum.deluge-torrent.org/viewtopic.php?f=7&t=55376&p=230417&hilit=magnet#p230417

Matching on torrent name appears to work with magnet links, but I don't think filename matching will work on them until Deluge 2.1.0 is released.

mhertz commented:
Thanks for reply :) The issue you're referrring to, is yarss2 plugin related and fixed in yarss2 v2.1.0, but I understand the confusion :)

I have tested that magnets work with autolabeling from filename, with your fork/pull-request. It only works however from gtk-UI and not consoleUI, and I haven't tested webUI. 

Right when the prefetching code was added to deluge, then original labelplus still worked, and I tested there if magnets worked for filename-autolabeling from consoleUI, but didn't, so then I fixed it, or workarounded it, with adding that small sleep() statement, to let metadata be resolved before labelplus did the matching, but which doesn't work anymore.

On the original labelplus github(well, here lol ;) ), there where also a pull-reqest you can see which adds support for sonarr etc, and which uses my patch, just ups the secs waiting alittle, as I also have begun doing myself, just doesn't work no more unfortunetly, which is little strange to me, as I can see the pause happens correctly, seemingly at right spot i.e. right when torrent added and before stating "torrent added" on terminal, e.g. when using 10 secs sleep. 

There are many download-tools that use deluge-console to add torrents to deluge, and if using public-torrents then pretty much all have changed to magnets nowadays, so imho an important issue missing.

Again, if you ever have any ideas that possibly could help me naroowing down how to implement this workaround again(where to use sleep() ), then it would be very much appreciated. And again, I apologize for off-topic'ness.

bdutro committed:
Properly handle magnet links
- Adds a handler for the metadata_received_alert that re-labels any
magnet torrents

bdutro commented:
@mhertz, thanks for the explanation. I added an additional handler that re-labels magnet torrents once their metadata is downloaded. Let me know if it works for you.

*Edit:* I uploaded new versions of the Python eggs to my release area as well.

mhertz commented:
Wow, I'm flabbergasted! Thank you so much for that, I really appreciate this alot! It works great and really nice to have a propper solution instead of my stupid workaround :) Thank you so much again for your efforts, you rock - not only making this plugin deluge2 compatible and then also fixing the magnet-issue! Thank you! :)

Untill bdutro makes a new release on his page, then here is the newest version with included magnet-fix: http://s000.tinyupload.com/index.php?file_id=03904783524220620978 (Possibly needing renamed to py3.6 instead of 3.7 if not being able to install it)

Edit: Nevermind above link, as bdutro already uploaded new release.

bdutro committed:
- Bump version number to 0.3.2.4

galfwender commented:
Hello, I am trying to use LabelPlus 0.3.2.4 in Deluge 2.0.3 to change the default download location for torrents added by Sonarr and Radarr.  It correctly applies a LabelPlus label to these torrents and sets other settings (e.g. download speed, move on completion), but it does not change the initial download location.  Any ideas?

mhertz commented:
bdutro, sorry for posting here, but your repo doesn't have any issues section enabled.

Anyway, on labelplus thread on deluge-forum, several recently stated issues with move_storage fails with your plugin, and when I tested this, then also it failed on me, but only on libtorrent 1.2.9 and not 1.2.8, which was all I tested, on arch-linux, with deluge 2.0.4-dev38 and a magnet link - it failed both move_storage and download-location-changing with libtorrent 1.2.9 in two different tests for me - here's one of the debug-log relevant parts and thanks in advance if having time at some point to possibly look into this here alittle, thanks:
```
20:45:01.968 [DEBUG   ][deluge.core.alertmanager      :135 ] Handling alert: torrent_checked_alert
20:45:01.972 [DEBUG   ][deluge.core.torrent           :1292] Writing torrent file to: /home/martin/.config/deluge/state/49da7ae0de8874462471d0f5419b850e599b05ef.torrent
20:45:01.973 [DEBUG   ][deluge.core.torrent           :511 ] Setting 49da7ae0de8874462471d0f5419b850e599b05ef file priorities to: []
20:45:01.973 [DEBUG   ][deluge.core.torrent           :518 ] Unable to set new file priorities.
20:45:01.973 [DEBUG   ][labelplus.core.core           :1635] [LabelPlus] Setting torrent '49da7ae0de8874462471d0f5419b850e599b05ef' to label '0'
20:45:01.974 [CRITICAL][twisted                       :154 ] Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/deluge/core/daemon_entry.py", line 122, in run_daemon
    daemon.start()
  File "/usr/lib/python3.8/site-packages/deluge/core/daemon.py", line 164, in start
    reactor.run()
  File "/usr/lib/python3.8/site-packages/twisted/internet/base.py", line 1283, in run
    self.mainLoop()
  File "/usr/lib/python3.8/site-packages/twisted/internet/base.py", line 1292, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python3.8/site-packages/twisted/internet/base.py", line 913, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/home/martin/.config/deluge/plugins/LabelPlus-0.3.2.4-py3.8.egg/labelplus/core/core.py", line 98, in wrap
   
  File "/home/martin/.config/deluge/plugins/LabelPlus-0.3.2.4-py3.8.egg/labelplus/core/core.py", line 752, in on_torrent_metadata_received
   
  File "/home/martin/.config/deluge/plugins/LabelPlus-0.3.2.4-py3.8.egg/labelplus/core/core.py", line 98, in wrap
   
  File "/home/martin/.config/deluge/plugins/LabelPlus-0.3.2.4-py3.8.egg/labelplus/core/core.py", line 736, in on_torrent_added
   
  File "/home/martin/.config/deluge/plugins/LabelPlus-0.3.2.4-py3.8.egg/labelplus/core/core.py", line 1704, in _move_torrents
   
  File "/usr/lib/python3.8/site-packages/deluge/core/torrent.py", line 1252, in move_storage
    self.handle.move_storage(dest, flags=2)
Boost.Python.ArgumentError: Python argument types in
    torrent_handle.move_storage(torrent_handle, str)
did not match C++ signature:
    move_storage(libtorrent::torrent_handle {lvalue}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > path, libtorrent::move_flags_t flags=libtorrent.move_flags_t.always_replace_files)
```

bdutro commented:
@mhertz sorry about that, I went ahead and enabled issues on my fork. If you don't mind, make an issue and we'll move the discussion over there.

mhertz commented:
Thanks for adding that issues section, much appreciated and will come in handy in the feature. However, i'm sorry bdutro, I just discovered that my issue here, where irrelevant to labelplus and was an issue in deluge directly, when used with libtorrent 1.2.9, likewise qbittorrent(-nox) will not even start with that version I just found out too. I will make an issues ticket if finding an actual issue with this, and/or direct others to do that from deluge-forum, as since stated, there have been some reports about this lately(unrelated to above issue, as reported before libtorrent 1.2.9 released). Sorry for wasting you time, and thanks for your prompt support and helpfullness, much appreciated!
